File: test/test_input_aligner.cpp
// Copyright 2024, Kraken Robotics. All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//    * Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//    * Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in the
//      documentation and/or other materials provided with the distribution.
//
//    * Neither the name of the Willow Garage nor the names of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#include <gtest/gtest.h>

#include <array>
#include <chrono>
#include <cstddef>
#include <memory>
#include <type_traits>
#include <vector>

#include <rclcpp/rclcpp.hpp>
#include "message_filters/input_aligner.hpp"

class InputAligner;

struct Header
{
  rclcpp::Time stamp;
};

struct Msg1
{
  Header header;
  int data;
};

struct Msg2
{
  Header header;
  int data;
};

namespace message_filters
{
namespace message_traits
{
template<typename M>
struct HasHeader<M, typename std::enable_if<std::is_same<Header,
  decltype(M().header)>::value>::type>: public std::true_type {};
}  // namespace message_traits
}  // namespace message_filters

class InputAlignerTest : public ::testing::Test
{
public:
  InputAlignerTest()
  : timeout_(1, 0), update_rate_(0, 1000000)
  {
  }

  virtual void SetUp()
  {
    // Shutdown in case there was a dangling global context from other test fixtures
    rclcpp::shutdown();
    rclcpp::init(0, nullptr);

    node_ = std::make_shared<rclcpp::Node>("test_node");
    timeout_ = rclcpp::Duration(1, 0);
    update_rate_ = rclcpp::Duration(0, 10000000);
  }

  void TearDown()
  {
    node_.reset();
    rclcpp::shutdown();
  }

  template<typename MsgType>
  void cb(const std::shared_ptr<const MsgType> & msg)
  {
    cb_content_.push_back(msg->data);
  }

  template<typename MsgType>
  std::shared_ptr<const MsgType> createMsg(const std::chrono::nanoseconds & ns, int data)
  {
    auto msg = std::make_shared<MsgType>();
    msg->header.stamp = rclcpp::Time(ns.count(), RCL_ROS_TIME);
    msg->data = data;
    return msg;
  }

  rclcpp::Node::SharedPtr node_;
  rclcpp::Duration timeout_;
  rclcpp::Duration update_rate_;
  std::vector<int> cb_content_;
};

TEST_F(InputAlignerTest, init)
{
  message_filters::NullFilter<Msg1> f0, f1;
  message_filters::NullFilter<Msg2> f2, f3;
  message_filters::InputAligner<Msg1, Msg1, Msg2, Msg2> aligner1(timeout_, f0, f1, f2, f3);

  message_filters::InputAligner<Msg1, Msg2, Msg2> aligner2(timeout_);
  aligner2.connectInput(f0, f2, f3);
}


TEST_F(InputAlignerTest, dispatch_inputs_in_order)
{
  message_filters::InputAligner<Msg1, Msg2, Msg1, Msg2> aligner(timeout_);

  // register callbacks
  InputAlignerTest * test_fixture = dynamic_cast<InputAlignerTest *>(this);
  aligner.registerCallback<0>(&InputAlignerTest::cb<Msg1>, test_fixture);
  aligner.registerCallback<1>(&InputAlignerTest::cb<Msg2>, test_fixture);
  aligner.registerCallback<2>(&InputAlignerTest::cb<Msg1>, test_fixture);
  aligner.registerCallback<3>(&InputAlignerTest::cb<Msg2>, test_fixture);

  // set period for all inputs to one millisecond
  aligner.setInputPeriod<0>(rclcpp::Duration(0, 4e6));
  aligner.setInputPeriod<1>(rclcpp::Duration(0, 4e6));
  aligner.setInputPeriod<2>(rclcpp::Duration(0, 4e6));
  aligner.setInputPeriod<3>(rclcpp::Duration(0, 4e6));

  // send msgs in unaligned order
  aligner.add<2>(createMsg<Msg1>(std::chrono::milliseconds(3), 3));
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(1), 1));
  aligner.add<2>(createMsg<Msg1>(std::chrono::milliseconds(7), 7));
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(5), 5));
  aligner.add<3>(createMsg<Msg2>(std::chrono::milliseconds(2), 2));
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(9), 9));
  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(4), 4));
  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(8), 8));
  aligner.add<3>(createMsg<Msg2>(std::chrono::milliseconds(6), 6));

  // dispatch messages
  aligner.dispatchMessages();

  // compare msg output
  ASSERT_EQ(cb_content_.size(), 9);
  for (size_t i = 0; i < cb_content_.size(); i++) {
    EXPECT_EQ(cb_content_[i], i + 1);
  }
}

TEST_F(InputAlignerTest, ignores_inactive_inputs)
{
  message_filters::InputAligner<Msg1, Msg2, Msg1> aligner(timeout_);

  // register callbacks
  InputAlignerTest * test_fixture = dynamic_cast<InputAlignerTest *>(this);
  aligner.registerCallback<0>(&InputAlignerTest::cb<Msg1>, test_fixture);
  aligner.registerCallback<1>(&InputAlignerTest::cb<Msg2>, test_fixture);
  aligner.registerCallback<2>(&InputAlignerTest::cb<Msg1>, test_fixture);

  // set period for all inputs to one millisecond
  aligner.setInputPeriod<0>(rclcpp::Duration(0, 2e6));
  aligner.setInputPeriod<1>(rclcpp::Duration(0, 2e6));
  aligner.setInputPeriod<2>(rclcpp::Duration(0, 2e6));

  // send msgs in unaligned order only to two inputs
  aligner.add<2>(createMsg<Msg1>(std::chrono::milliseconds(2), 2));
  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(1), 1));
  aligner.add<2>(createMsg<Msg1>(std::chrono::milliseconds(4), 4));
  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(3), 3));
  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(5), 5));

  // dispatch messages
  aligner.dispatchMessages();

  // compare msg output
  ASSERT_EQ(cb_content_.size(), 5);
  for (size_t i = 0; i < cb_content_.size(); i++) {
    EXPECT_EQ(cb_content_[i], i + 1);
  }
}

TEST_F(InputAlignerTest, input_timeout)
{
  // set timeout of 10 ms
  timeout_ = rclcpp::Duration(0, 1e7);
  message_filters::InputAligner<Msg1, Msg2> aligner(timeout_);

  // register callbacks
  InputAlignerTest * test_fixture = dynamic_cast<InputAlignerTest *>(this);
  aligner.registerCallback<0>(&InputAlignerTest::cb<Msg1>, test_fixture);
  aligner.registerCallback<1>(&InputAlignerTest::cb<Msg2>, test_fixture);

  // set period for all inputs to 2 milliseconds
  aligner.setInputPeriod<0>(rclcpp::Duration(0, 2e6));
  aligner.setInputPeriod<1>(rclcpp::Duration(0, 2e6));

  // adds 8 msgs on input 0
  for (int i = 1; i < 17; i = i + 2) {
    aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(i), i));
  }
  // adds 2 msgs on input 1
  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(2), 2));
  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(4), 4));

  // dispatch messages
  aligner.dispatchMessages();

  // compare msg output
  // only msgs < 6 ms are forwarded, it will wait for msgs on input 1
  ASSERT_EQ(cb_content_.size(), 5);
  for (size_t i = 0; i < cb_content_.size(); i++) {
    EXPECT_EQ(cb_content_[i], i + 1);
  }

  // a new msg with a time stamp > 10 ms from the
  // last forwarded one will trigger the timeout for input 1
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(17), 17));

  // dispatch messages
  aligner.dispatchMessages();
  ASSERT_EQ(cb_content_.size(), 11);
  EXPECT_EQ(cb_content_[8], 13);
  EXPECT_EQ(cb_content_[9], 15);
  EXPECT_EQ(cb_content_[10], 17);
}

TEST_F(InputAlignerTest, drops_msgs)
{
  message_filters::InputAligner<Msg1, Msg2> aligner(timeout_);

  // register callbacks
  InputAlignerTest * test_fixture = dynamic_cast<InputAlignerTest *>(this);
  aligner.registerCallback<0>(&InputAlignerTest::cb<Msg1>, test_fixture);
  aligner.registerCallback<1>(&InputAlignerTest::cb<Msg2>, test_fixture);

  // set period for all inputs to 2 milliseconds
  aligner.setInputPeriod<0>(rclcpp::Duration(0, 2e6));
  aligner.setInputPeriod<1>(rclcpp::Duration(0, 2e6));


  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(4), 4));
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(3), 3));

  // dispatch messages
  aligner.dispatchMessages();

  ASSERT_EQ(cb_content_.size(), 2);
  for (size_t i = 0; i < cb_content_.size(); i++) {
    EXPECT_EQ(cb_content_[i], i + 3);
  }

  // add also msg backwards in time
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(1), 1));
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(5), 5));
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(7), 7));
  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(2), 2));
  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(6), 6));

  // dispatch messages
  aligner.dispatchMessages();

  ASSERT_EQ(cb_content_.size(), 5);
  for (size_t i = 0; i < cb_content_.size(); i++) {
    EXPECT_EQ(cb_content_[i], i + 3);
  }
}

TEST_F(InputAlignerTest, dispatch_by_timer)
{
  message_filters::InputAligner<Msg1, Msg2> aligner(timeout_);
  aligner.setupDispatchTimer(node_, update_rate_);

  // register callbacks
  InputAlignerTest * test_fixture = dynamic_cast<InputAlignerTest *>(this);
  aligner.registerCallback<0>(&InputAlignerTest::cb<Msg1>, test_fixture);
  aligner.registerCallback<1>(&InputAlignerTest::cb<Msg2>, test_fixture);

  // set period for all inputs to 2 milliseconds
  aligner.setInputPeriod<0>(rclcpp::Duration(0, 2e6));
  aligner.setInputPeriod<1>(rclcpp::Duration(0, 2e6));

  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(2), 2));
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(1), 1));

  rclcpp::Rate(50).sleep();
  rclcpp::spin_some(node_);

  ASSERT_EQ(cb_content_.size(), 2);
  for (size_t i = 0; i < cb_content_.size(); i++) {
    EXPECT_EQ(cb_content_[i], i + 1);
  }
}

TEST_F(InputAlignerTest, no_period_information)
{
  // set timeout of 10 ms
  timeout_ = rclcpp::Duration(0, 1e7);
  message_filters::InputAligner<Msg1, Msg2, Msg1> aligner(timeout_);

  // register callbacks
  InputAlignerTest * test_fixture = dynamic_cast<InputAlignerTest *>(this);
  aligner.registerCallback<0>(&InputAlignerTest::cb<Msg1>, test_fixture);
  aligner.registerCallback<1>(&InputAlignerTest::cb<Msg2>, test_fixture);
  aligner.registerCallback<2>(&InputAlignerTest::cb<Msg1>, test_fixture);

  // send msgs in unaligned order only to two inputs
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(6), 6));
  aligner.add<2>(createMsg<Msg1>(std::chrono::milliseconds(2), 2));
  aligner.add<2>(createMsg<Msg1>(std::chrono::milliseconds(4), 4));
  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(1), 1));
  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(3), 3));
  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(5), 5));

  // dispatch messages
  aligner.dispatchMessages();

  // the last two messages won't get dispatched
  ASSERT_EQ(cb_content_.size(), 4);
  for (size_t i = 0; i < cb_content_.size(); i++) {
    EXPECT_EQ(cb_content_[i], i + 1);
  }

  // trigger timeout on all other inputs to flush msgs
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(16), 16));

  // dispatch messages
  aligner.dispatchMessages();
  ASSERT_EQ(cb_content_.size(), 7);
}

TEST_F(InputAlignerTest, get_queue_status)
{
  // set timeout of 10 ms
  timeout_ = rclcpp::Duration(0, 1e7);
  message_filters::InputAligner<Msg1, Msg2> aligner(timeout_);

  // register callbacks
  InputAlignerTest * test_fixture = dynamic_cast<InputAlignerTest *>(this);
  aligner.registerCallback<0>(&InputAlignerTest::cb<Msg1>, test_fixture);
  aligner.registerCallback<1>(&InputAlignerTest::cb<Msg2>, test_fixture);

  // set period for all inputs to 2 milliseconds
  aligner.setInputPeriod<0>(rclcpp::Duration(0, 2e6));
  aligner.setInputPeriod<1>(rclcpp::Duration(0, 2e6));

  aligner.add<1>(createMsg<Msg2>(std::chrono::milliseconds(2), 2));
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(3), 3));
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(5), 5));

  auto status_0 = aligner.getQueueStatus<0>();
  EXPECT_FALSE(status_0.active);
  EXPECT_EQ(status_0.queue_size, 2);
  EXPECT_EQ(status_0.msgs_processed, 0);
  EXPECT_EQ(status_0.msgs_dropped, 0);
  auto status_1 = aligner.getQueueStatus<1>();
  EXPECT_FALSE(status_1.active);
  EXPECT_EQ(status_1.queue_size, 1);
  EXPECT_EQ(status_1.msgs_processed, 0);
  EXPECT_EQ(status_1.msgs_dropped, 0);

  // dispatch messages
  aligner.dispatchMessages();

  status_0 = aligner.getQueueStatus<0>();
  EXPECT_TRUE(status_0.active);
  EXPECT_EQ(status_0.queue_size, 1);
  EXPECT_EQ(status_0.msgs_processed, 1);
  EXPECT_EQ(status_0.msgs_dropped, 0);
  status_1 = aligner.getQueueStatus<1>();
  EXPECT_TRUE(status_1.active);
  EXPECT_EQ(status_1.queue_size, 0);
  EXPECT_EQ(status_1.msgs_processed, 1);
  EXPECT_EQ(status_1.msgs_dropped, 0);

  // adds sample outside of filter scope
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(1), 1));
  // adds a sample ahead forcing input 1 to time out
  aligner.add<0>(createMsg<Msg1>(std::chrono::milliseconds(17), 17));

  // dispatch messages
  aligner.dispatchMessages();

  status_0 = aligner.getQueueStatus<0>();
  EXPECT_TRUE(status_0.active);
  EXPECT_EQ(status_0.queue_size, 0);
  EXPECT_EQ(status_0.msgs_processed, 3);
  EXPECT_EQ(status_0.msgs_dropped, 1);
  status_1 = aligner.getQueueStatus<1>();
  EXPECT_FALSE(status_1.active);
  EXPECT_EQ(status_1.queue_size, 0);
  EXPECT_EQ(status_1.msgs_processed, 1);
  EXPECT_EQ(status_1.msgs_dropped, 0);
}

File: test/test_input_aligner.py
# Copyright 2026, Open Source Robotics Foundation, Inc. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#    * Redistributions of source code must retain the above copyright
#      notice, this list of conditions and the following disclaimer.
#
#    * Redistributions in binary form must reproduce the above copyright
#      notice, this list of conditions and the following disclaimer in the
#      documentation and/or other materials provided with the distribution.
#
#    * Neither the name of the Willow Garage nor the names of its
#      contributors may be used to endorse or promote products derived from
#      this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.


import time
import unittest

from builtin_interfaces.msg import Time as TimeMsg
from message_filters import SimpleFilter
from message_filters.input_aligner import InputAligner
import rclpy
from rclpy.duration import Duration
from rclpy.time import Time


class Header:

    def __init__(self, stamp=None):
        self.stamp = stamp if stamp is not None else TimeMsg()


class Msg1:

    def __init__(self, stamp=None, data=None):
        self.header = Header(stamp)
        self.data = data


class Msg2:

    def __init__(self, stamp=None, data=None):
        self.header = Header(stamp)
        self.data = data


class TestInputAligner(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        rclpy.init()
        cls.node = rclpy.create_node('test_input_aligner_node')

    @classmethod
    def tearDownClass(cls):
        cls.node.destroy_node()
        rclpy.shutdown()

    def setUp(self):
        self.timeout = Duration(seconds=1.0)
        self.update_rate = Duration(nanoseconds=10000000)
        self.callback_content = []

    def callback(self, msg):
        self.callback_content.append(msg.data)

    def create_msg(self, cls, milliseconds, data):
        return cls(stamp=Time(nanoseconds=int(milliseconds * 1e6)).to_msg(), data=data)

    def test_init(self):
        f0, f1, f2, f3 = SimpleFilter(), SimpleFilter(), SimpleFilter(), SimpleFilter()
        aligner1 = InputAligner(self.timeout, filters=[f0, f1, f2, f3])
        self.assertEqual(len(aligner1.event_queues), 4)
        aligner2 = InputAligner(self.timeout)
        aligner2.connectInput(filters=[f0, f2, f3])
        self.assertEqual(len(aligner2.event_queues), 3)

    def test_set_and_get_name(self):
        aligner = InputAligner(self.timeout)
        self.assertEqual(aligner.getName(), '')
        aligner.setName('camera_aligner')
        self.assertEqual(aligner.getName(), 'camera_aligner')

    def test_dispatch_inputs_in_order(self):
        aligner = InputAligner(self.timeout)
        aligner.connectInput(
            filters=[SimpleFilter(), SimpleFilter(), SimpleFilter(), SimpleFilter()])
        for i in range(4):
            aligner.registerCallback(i, self.callback)
            aligner.setInputPeriod(i, Duration(nanoseconds=int(4e6)))
        aligner.add(self.create_msg(Msg1, 3, 3), 2)
        aligner.add(self.create_msg(Msg1, 1, 1), 0)
        aligner.add(self.create_msg(Msg1, 7, 7), 2)
        aligner.add(self.create_msg(Msg1, 5, 5), 0)
        aligner.add(self.create_msg(Msg2, 2, 2), 3)
        aligner.add(self.create_msg(Msg1, 9, 9), 0)
        aligner.add(self.create_msg(Msg2, 4, 4), 1)
        aligner.add(self.create_msg(Msg2, 8, 8), 1)
        aligner.add(self.create_msg(Msg2, 6, 6), 3)
        aligner.dispatchMessages()
        self.assertEqual(self.callback_content, list(range(1, 10)))

    def test_dispatch_inputs_with_duplicate_timestamps(self):
        aligner = InputAligner(self.timeout)
        aligner.connectInput(
            filters=[SimpleFilter(), SimpleFilter(), SimpleFilter(), SimpleFilter()])
        for i in range(4):
            aligner.registerCallback(i, self.callback)
            aligner.setInputPeriod(i, Duration(nanoseconds=int(4e6)))
        aligner.add(self.create_msg(Msg1, 3, 3), 2)
        aligner.add(self.create_msg(Msg1, 1, 1), 0)
        aligner.add(self.create_msg(Msg1, 7, 7), 2)
        aligner.add(self.create_msg(Msg1, 5, 5), 0)
        aligner.add(self.create_msg(Msg2, 2, 2), 3)
        aligner.add(self.create_msg(Msg1, 9, 9), 0)
        # Two messages with identical timestamp on the same queue should
        # dispatch in insertion order via the seq_id tie-breaker.
        aligner.add(self.create_msg(Msg1, 9, 9), 0)
        aligner.add(self.create_msg(Msg2, 4, 4), 1)
        aligner.add(self.create_msg(Msg2, 8, 8), 1)
        aligner.add(self.create_msg(Msg2, 6, 6), 3)
        aligner.dispatchMessages()
        self.assertEqual(self.callback_content, [1, 2, 3, 4, 5, 6, 7, 8, 9, 9])

    def test_reconnect_input_disconnects_old_upstream(self):
        f0, f1, f2 = SimpleFilter(), SimpleFilter(), SimpleFilter()
        aligner = InputAligner(self.timeout)
        aligner.connectInput(filters=[f0, f1, f2])
        aligner.connectInput(filters=[f0, f1])
        for i in range(2):
            aligner.registerCallback(i, self.callback)
            aligner.setInputPeriod(i, Duration(nanoseconds=int(2e6)))
        # f2 was dropped on reconnect; messages from it must be ignored.
        f2.signalMessage(self.create_msg(Msg1, 1, 1))
        f0.signalMessage(self.create_msg(Msg1, 2, 2))
        f1.signalMessage(self.create_msg(Msg2, 3, 3))
        aligner.dispatchMessages()
        self.assertEqual(self.callback_content, [2, 3])

    def test_reconnect_input_drops_old_downstream_callbacks(self):
        # Callbacks registered before reconnect should be discarded along with
        # the per-input signals they were attached to.
        stale = []
        fresh = []
        f0, f1 = SimpleFilter(), SimpleFilter()
        aligner = InputAligner(self.timeout)
        aligner.connectInput(filters=[f0, f1])
        aligner.registerCallback(0, lambda m: stale.append(m.data))
        aligner.registerCallback(1, lambda m: stale.append(m.data))
        aligner.connectInput(filters=[f0, f1])
        aligner.registerCallback(0, lambda m: fresh.append(m.data))
        aligner.registerCallback(1, lambda m: fresh.append(m.data))
        for i in range(2):
            aligner.setInputPeriod(i, Duration(nanoseconds=int(2e6)))
        aligner.add(self.create_msg(Msg1, 1, 1), 0)
        aligner.add(self.create_msg(Msg2, 2, 2), 1)
        aligner.dispatchMessages()
        self.assertEqual(stale, [])
        self.assertEqual(fresh, [1, 2])

    def test_disconnect_all_stops_upstream_delivery(self):
        f0, f1 = SimpleFilter(), SimpleFilter()
        aligner = InputAligner(self.timeout)
        aligner.connectInput(filters=[f0, f1])
        for i in range(2):
            aligner.registerCallback(i, self.callback)
            aligner.setInputPeriod(i, Duration(nanoseconds=int(2e6)))
        aligner.disconnectAll()
        f0.signalMessage(self.create_msg(Msg1, 1, 1))
        f1.signalMessage(self.create_msg(Msg2, 2, 2))
        aligner.dispatchMessages()
        self.assertEqual(self.callback_content, [])

    def test_simple_filter_unregister_callback(self):
        received = []
        f = SimpleFilter()
        conn_a = f.registerCallback(lambda m: received.append(('a', m)))
        f.registerCallback(lambda m: received.append(('b', m)))
        f.signalMessage(1)
        self.assertEqual(received, [('a', 1), ('b', 1)])
        f.unregisterCallback(conn_a)
        f.signalMessage(2)
        self.assertEqual(received, [('a', 1), ('b', 1), ('b', 2)])
        # Unregistering an unknown id must be a no-op.
        f.unregisterCallback(conn_a)
        f.unregisterCallback(9999)
        f.signalMessage(3)
        self.assertEqual(received, [('a', 1), ('b', 1), ('b', 2), ('b', 3)])

    def test_ignores_inactive_inputs(self):
        aligner = InputAligner(self.timeout)
        aligner.connectInput(filters=[SimpleFilter(), SimpleFilter(), SimpleFilter()])
        for i in range(3):
            aligner.registerCallback(i, self.callback)
            aligner.setInputPeriod(i, Duration(nanoseconds=int(2e6)))
        aligner.add(self.create_msg(Msg1, 2, 2), 2)
        aligner.add(self.create_msg(Msg2, 1, 1), 1)
        aligner.add(self.create_msg(Msg1, 4, 4), 2)
        aligner.add(self.create_msg(Msg2, 3, 3), 1)
        aligner.add(self.create_msg(Msg2, 5, 5), 1)
        aligner.dispatchMessages()
        self.assertEqual(self.callback_content, [1, 2, 3, 4, 5])

    def test_input_timeout(self):
        self.timeout = Duration(nanoseconds=int(1e7))
        aligner = InputAligner(self.timeout)
        aligner.connectInput(filters=[SimpleFilter(), SimpleFilter()])
        for i in range(2):
            aligner.registerCallback(i, self.callback)
            aligner.setInputPeriod(i, Duration(nanoseconds=int(2e6)))
        for i in range(1, 17, 2):
            aligner.add(self.create_msg(Msg1, i, i), 0)
        aligner.add(self.create_msg(Msg2, 2, 2), 1)
        aligner.add(self.create_msg(Msg2, 4, 4), 1)
        aligner.dispatchMessages()
        self.assertEqual(self.callback_content, [1, 2, 3, 4, 5])
        aligner.add(self.create_msg(Msg1, 17, 17), 0)
        aligner.dispatchMessages()
        self.assertEqual(self.callback_content, [1, 2, 3, 4, 5, 7, 9, 11, 13, 15, 17])

    def test_drops_msgs(self):
        aligner = InputAligner(self.timeout)
        aligner.connectInput(filters=[SimpleFilter(), SimpleFilter()])
        for i in range(2):
            aligner.registerCallback(i, self.callback)
            aligner.setInputPeriod(i, Duration(nanoseconds=int(2e6)))
        aligner.add(self.create_msg(Msg2, 4, 4), 1)
        aligner.add(self.create_msg(Msg1, 3, 3), 0)
        aligner.dispatchMessages()
        self.assertEqual(self.callback_content, [3, 4])
        aligner.add(self.create_msg(Msg1, 1, 1), 0)
        aligner.add(self.create_msg(Msg1, 5, 5), 0)
        aligner.add(self.create_msg(Msg1, 7, 7), 0)
        aligner.add(self.create_msg(Msg2, 2, 2), 1)
        aligner.add(self.create_msg(Msg2, 6, 6), 1)
        aligner.dispatchMessages()
        self.assertEqual(self.callback_content, [3, 4, 5, 6, 7])

    def test_dispatch_by_timer(self):
        aligner = InputAligner(self.timeout)
        aligner.connectInput(filters=[SimpleFilter(), SimpleFilter()])
        aligner.setupDispatchTimer(self.node, self.update_rate)
        for i in range(2):
            aligner.registerCallback(i, self.callback)
            aligner.setInputPeriod(i, Duration(nanoseconds=int(2e6)))
        aligner.add(self.create_msg(Msg2, 2, 2), 1)
        aligner.add(self.create_msg(Msg1, 1, 1), 0)
        # Spin until the timer fires and both messages flow through, with a
        # generous deadline so this is not flaky on slow CI runners.
        deadline = time.monotonic() + 2.0
        while len(self.callback_content) < 2 and time.monotonic() < deadline:
            rclpy.spin_once(self.node, timeout_sec=0.05)
        self.assertEqual(self.callback_content, [1, 2])

    def test_no_period_information(self):
        self.timeout = Duration(nanoseconds=int(1e7))
        aligner = InputAligner(self.timeout)
        aligner.connectInput(filters=[SimpleFilter(), SimpleFilter(), SimpleFilter()])
        for i in range(3):
            aligner.registerCallback(i, self.callback)
        aligner.add(self.create_msg(Msg1, 6, 6), 0)
        aligner.add(self.create_msg(Msg1, 2, 2), 2)
        aligner.add(self.create_msg(Msg1, 4, 4), 2)
        aligner.add(self.create_msg(Msg2, 1, 1), 1)
        aligner.add(self.create_msg(Msg2, 3, 3), 1)
        aligner.add(self.create_msg(Msg2, 5, 5), 1)
        aligner.dispatchMessages()
        self.assertEqual(self.callback_content, [1, 2, 3, 4])
        aligner.add(self.create_msg(Msg1, 16, 16), 0)
        aligner.dispatchMessages()
        self.assertEqual(self.callback_content, [1, 2, 3, 4, 5, 6, 16])

    def test_get_queue_status(self):
        self.timeout = Duration(nanoseconds=int(1e7))
        aligner = InputAligner(self.timeout)
        aligner.connectInput(filters=[SimpleFilter(), SimpleFilter()])
        for i in range(2):
            aligner.registerCallback(i, self.callback)
            aligner.setInputPeriod(i, Duration(nanoseconds=int(2e6)))
        aligner.add(self.create_msg(Msg2, 2, 2), 1)
        aligner.add(self.create_msg(Msg1, 3, 3), 0)
        aligner.add(self.create_msg(Msg1, 5, 5), 0)
        status_0 = aligner.getQueueStatus(0)
        self.assertFalse(status_0.active)
        self.assertEqual(status_0.queue_size, 2)
        self.assertEqual(status_0.msgs_processed, 0)
        self.assertEqual(status_0.msgs_dropped, 0)
        status_1 = aligner.getQueueStatus(1)
        self.assertFalse(status_1.active)
        self.assertEqual(status_1.queue_size, 1)
        self.assertEqual(status_1.msgs_processed, 0)
        self.assertEqual(status_1.msgs_dropped, 0)
        aligner.dispatchMessages()
        status_0 = aligner.getQueueStatus(0)
        self.assertTrue(status_0.active)
        self.assertEqual(status_0.queue_size, 1)
        self.assertEqual(status_0.msgs_processed, 1)
        self.assertEqual(status_0.msgs_dropped, 0)
        status_1 = aligner.getQueueStatus(1)
        self.assertTrue(status_1.active)
        self.assertEqual(status_1.queue_size, 0)
        self.assertEqual(status_1.msgs_processed, 1)
        self.assertEqual(status_1.msgs_dropped, 0)
        aligner.add(self.create_msg(Msg1, 1, 1), 0)
        aligner.add(self.create_msg(Msg1, 17, 17), 0)
        aligner.dispatchMessages()
        status_0 = aligner.getQueueStatus(0)
        self.assertTrue(status_0.active)
        self.assertEqual(status_0.queue_size, 0)
        self.assertEqual(status_0.msgs_processed, 3)
        self.assertEqual(status_0.msgs_dropped, 1)
        status_1 = aligner.getQueueStatus(1)
        self.assertFalse(status_1.active)
        self.assertEqual(status_1.queue_size, 0)
        self.assertEqual(status_1.msgs_processed, 1)
        self.assertEqual(status_1.msgs_dropped, 0)


if __name__ == '__main__':
    unittest.main()
